lib/modules/coverage/index.js

```diff
@@ -0,0 +1,70 @@
+/*global web3*/
+const process = require('process');
+
+const fs = require('../../core/fs');
+const ContractSources = require('./contract_sources');
+
+// Set up the web3 extension
+web3.extend({
+  property: 'debug',
+  methods: [{name: 'traceTransaction', call: 'debug_traceTransaction', params: 2}]
+});
+
+class CodeCoverage {
+  constructor(embark, _options) {
+    this.events = embark.events;
+    this.logger = embark.logger;
+
+    embark.events.on('contracts:compile:solc', this.compileSolc.bind(this));
+    embark.events.on('contracts:compiled:solc', this.compiledSolc.bind(this));
+    embark.events.on('contracts:run:solc', this.runSolc.bind(this));
+    embark.events.on('block:header', this.runSolc.bind(this));
+
+    this.seenTransactions = {};
+
+    var self = this;
+
+    process.on('exit', (_code) => {
+      const coverageReportPath = fs.dappPath('.embark', 'coverage.json');
+      fs.writeFileSync(coverageReportPath, JSON.stringify(self.coverageReport));
```

Comment:
Wouldn't this create a race condition where your file write might not finish before the exit?

Reply:
Not if it's sync. It'll wait for that to finish.